File: Concepts/LinkedList/linkedlist.py
class Node:
    def __init__(self, data:int):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self) -> None:
        self.head = None
        self.tail = None
        self.length = 0 

    
    def append(self, n:Node) -> None:
        if self.head is None:
            self.head = n
            self.tail = n
        else:
            self.tail.next = n
            self.tail = n
        self.length += 1


    def prepend(self, n:Node) -> None:
        if self.head is None:
            self.head =  n
            self.tail =  n
        else:
            n.next = self.head
            self.head = n

    def print_list(self) -> None:
        to_print = self.head
        while to_print is not None:
            print(to_print.data, end=" ")
            to_print = to_print.next
        print()



    def delete_with_value(self, value:int) -> None:
        if self.head is None:
            print("No linked list found")
            return

        if self.head.data == value:
            self.head = self.head.next
            self.length -= 1
            return

        prev_to_delete = self.head

        while prev_to_delete.next is not None and prev_to_delete.next.data != value:
            prev_to_delete = prev_to_delete.next

        if prev_to_delete.next is None:
            print(f"No node found with data {value}")
            return

        prev_to_delete.next = prev_to_delete.next.next
        self.length -= 1


def main():
    list1 = LinkedList()

    node1 = Node(10)
    node2 = Node(11)
    node3 = Node(12)

    list1.prepend(node1)
    list1.prepend(node2)
    list1.prepend(node3)

    node4 = Node(13)
    node5 = Node(14)
    node6 = Node(15)

    list1.append(node4)
    list1.append(node5)
    list1.append(node6)

    list1.print_list()  # Expected: 12 11 10 13 14 15

    list1.delete_with_value(12)
    list1.print_list()  # Expected: 11 10 13 14 15

    list1.delete_with_value(120)
    list1.print_list()  # Expected: 11 10 13 14 15


if __name__ == "__main__":
    main()        

File: Concepts/LinkedList/linkedlist.go
package main

import "fmt"

type node struct {
	data int
	next *node
}

type linkedList struct {
	head   *node
	tail   *node
	length int
}

func (l *linkedList) prepend(n *node) {
	// If the list is empty, both head and tail point to the new node

	if l.head == nil {
		l.head = n
		l.tail = n
	} else {

		n.next = l.head
		l.head = n
	}

	// second := l.head

	// l.head = n

	// l.head.next = second

	l.length++

}

func (l linkedList) pritnList() {

	toPrint := l.head

	for toPrint != nil {

		fmt.Printf("%v ", toPrint.data)
		toPrint = toPrint.next
	}
	fmt.Println()

}

func (l *linkedList) append(n *node) {

	if l.head == nil {
		// If the list is empty, both head and tail point to the new node
		l.head = n
		l.tail = n
	} else {

		l.tail.next = n
		l.tail = n
	}

	l.length++

}

func (l *linkedList) deleteWithValue(value int) {

	prevToDelete := l.head

	// handle deletion of head node
	if l.head.data == value {
		l.head = l.head.next
		l.length--
		return
	}

	// handle deletion single node linked list``
	if l.length == 0 {
		fmt.Println(" No linked list found")
		return

	}

	for prevToDelete.next.data != value {

		prevToDelete = prevToDelete.next

		if prevToDelete.next == nil {
			fmt.Printf("no node found with data %v \n", value)
			return
		}

	}

	prevToDelete.next = prevToDelete.next.next

}

func main() {

	list1 := linkedList{}

	node1 := node{data: 10}
	node2 := node{data: 11}
	node3 := node{data: 12}

	list1.prepend(&node1)
	list1.prepend(&node2)
	list1.prepend(&node3)

	node4 := node{data: 13}
	node5 := node{data: 14}
	node6 := node{data: 15}

	list1.append(&node4)
	list1.append(&node5)
	list1.append(&node6)

	list1.pritnList()

	list1.deleteWithValue(12)

	list1.pritnList()

	list1.deleteWithValue(120)

	list1.pritnList()

}
